src/main/java/org/prebid/server/analytics/reporter/greenbids/GreenbidsAnalyticsReporter.java

```diff
@@ -0,0 +1,355 @@
+
+    private static SeatNonBid toSeatNonBid(String bidder, BidRejectionTracker bidRejectionTracker) {
+        final List<NonBid> nonBids = bidRejectionTracker.getRejectionReasons().entrySet().stream()
+                .map(entry -> NonBid.of(entry.getKey(), entry.getValue()))
+                .toList();
+
+        return SeatNonBid.of(bidder, nonBids);
+    }
+
+    private List<AdUnit> extractAdUnitsWithBidResponses(
+            List<Imp> imps,
+            Map<String, Bid> seatsWithBids,
+            Map<String, NonBid> seatsWithNonBids) {
+        return imps.stream().map(imp -> createAdUnit(
+                imp, seatsWithBids, seatsWithNonBids)).toList();
+    }
+
+    private AdUnit createAdUnit(
+            Imp imp,
+            Map<String, Bid> seatsWithBids,
+            Map<String, NonBid> seatsWithNonBids) {
+        final Banner banner = imp.getBanner();
+        final Video video = imp.getVideo();
+        final Native nativeObject = imp.getXNative();
+
+        if (banner == null) {
+            Future.failedFuture(new PreBidException("Error: Banner should be non-null"));
+        }
+
+        // Error is here
+        final List<List<Integer>> bannerWidthHeight = Optional.ofNullable(banner)
+                .map(Banner::getFormat)
+                .map(formats -> formats.stream().map(format -> List.of(format.getW(), format.getH())))
+                .orElse(null).collect(Collectors.toList());
+
+        final ExtBanner extBanner = ExtBanner.builder()
+                .sizes(bannerWidthHeight)
+                .pos(Optional.ofNullable(banner).map(Banner::getPos).orElse(null))
+                .name(Optional.ofNullable(banner).map(Banner::getId).orElse(null))
+                .build();
+
+        final MediaTypes mediaTypes = MediaTypes.builder()
+                .banner(extBanner)
+                .video(video)
+                .nativeObject(nativeObject)
+                .build();
+
+        final List<GreenbidsBids> bids = extractBidders(imp, seatsWithBids, seatsWithNonBids);
+
+        final String adUnitCode = getAdUnitCode(imp);
+
+        return AdUnit.builder()
+                .code(adUnitCode)
+                .mediaTypes(mediaTypes)
+                .bids(bids)
+                .build();
+    }
+
+    private List<GreenbidsBids> extractBidders(
+            Imp imp, Map<String, Bid> seatsWithBids, Map<String, NonBid> seatsWithNonBids) {
+        final List<GreenbidsBids> bidders = new ArrayList<>();
+
+        final Map<String, Bid> seatsWithBidsForImp = seatsWithBids.entrySet().stream()
+                .filter(entry -> entry.getValue().getImpid().equals(imp.getId()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        final Map<String, NonBid> seatsWithNonBidsForImp = seatsWithNonBids.entrySet().stream()
+                .filter(entry -> entry.getValue().getImpId().equals(imp.getId()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
```

Comment:
is it possible to apply this filter earlier, when you build the `seatsWithBids` and `seatsWithNonBids`?

Reply:
and you can do `foreach` in the single chain without redunct collecting to map

Reply:
Have used `Stream.concat` to collect both streams of GreenbidsBids (for bids and nonBids) into a single list without intermediate collecting to map 👍 

Concerning the filter `.filter(entry -> entry.getValue().getImpid().equals(imp.getId()))` I'm not sure because if we apply it earlier in `createAdUnit` or in `imps.stream().map(imp -> createAdUnit())` we would still need to collect back to maps after filtering. So I propose to keep the filter in `extractBidders`